crates/matrix-sdk-crypto/src/store/mod.rs

```diff
@@ -204,15 +200,27 @@ impl<'a> SyncedKeyQueryManager<'a> {
         user: &UserId,
     ) -> UserKeyQueryResult {
         let mut users_for_key_query = self.manager.users_for_key_query.lock().await;
-
         let Some(waiter) = users_for_key_query.maybe_register_waiting_task(user) else {
             return UserKeyQueryResult::WasNotPending;
         };
 
         let wait_for_completion = async {
-            while !waiter.completed.load(Ordering::Relaxed) {
-                users_for_key_query =
-                    self.manager.users_for_key_query_condvar.wait(users_for_key_query).await;
+            loop {
+                // Register for being notified before releasing the mutex, so
+                // it's impossible to miss a wakeup between the last check for
+                // whether we should wait, and starting to wait.
+                let mut notified = pin!(self.manager.users_for_key_query_notify.notified());
+                notified.as_mut().enable();
+                drop(users_for_key_query);
+
+                // Wait for a notification
+                notified.await;
+
+                if waiter.completed.load(Ordering::Relaxed) {
+                    break;
+                } else {
+                    users_for_key_query = self.manager.users_for_key_query.lock().await;
+                }
```

Comment:
I think there's still a race here? the update might arrive between the call to `waiter.completed.load` and the reclaim of the lock. 

I think it might work if you reclaim the lock before the call to `waiter.completed.load`.

Reply:
Right, or with another check for `waiter.completed` at the start of the loop (kind of like it was before).